server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ShardsAllocator.java

```diff
@@ -33,22 +33,24 @@ public interface ShardsAllocator {
      * - relocate shards to find a good shard balance in the cluster
      *
      * @param allocation current node allocation
+     * @param listener listener to be executed once async allocation is completed
      */
-    void allocate(RoutingAllocation allocation);
+    void allocate(RoutingAllocation allocation, ActionListener<Void> listener);
 
     /**
-     * Allocates shards to nodes in the cluster. An implementation of this method should:
-     * - assign unassigned shards
-     * - relocate shards that cannot stay on a node anymore
-     * - relocate shards to find a good shard balance in the cluster
+     * Returns the decision for where a shard should reside in the cluster.  If the shard is unassigned,
+     * then the {@link AllocateUnassignedDecision} will be non-null.  If the shard is not in the unassigned
+     * state, then the {@link MoveDecision} will be non-null.
      *
-     * @param allocation current node allocation
-     * @param listener listener to be executed once async allocation is completed
+     * This method is primarily used by the cluster allocation explain API to provide detailed explanations
+     * for the allocation of a single shard.  Implementations of the {@link #allocate(RoutingAllocation, ActionListener<Void>)} method
+     * may use the results of this method implementation to decide on allocating shards in the routing table
+     * to the cluster.
+     *
+     * If an implementation of this interface does not support explaining decisions for a single shard through
+     * the cluster explain API, then this method should throw a {@code UnsupportedOperationException}.
      */
-    default void allocate(RoutingAllocation allocation, ActionListener<Void> listener) {
-        allocate(allocation);
-        listener.onResponse(null);
-    }
```

Comment:
This is the only thing removed. Diff is caused by moving default `execute` method to the bottomo.